### PR TITLE
Add notes about arity and file system types

### LINKED ISSUE #28706: Add example for FileSystemInfo
* Add FileSystemInfo example
* Explain how arity is determined by default
* Explain that required option doesn't have to be provided on command line if it has a default value

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 75416731-e081-9d04-e683-e0fba0d6b388
* Version Independent ID: 1eec00b7-55ba-34ce-225e-3549b9de4b25
* Content: [How to bind arguments to handlers in System.CommandLine](https://docs.microsoft.com/en-us/dotnet/standard/commandline/model-binding)
* Content Source: [docs/standard/commandline/model-binding.md](https://github.com/dotnet/docs/blob/main/docs/standard/commandline/model-binding.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @gewarren
* Microsoft Alias: **gewarren**

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #28706

### CODE REVIEW

File: docs/standard/commandline/model-binding.md [hunk 1]
```diff
@@ -149,13 +149,17 @@ Because `AllowMultipleArgumentsPerToken` is set to `true`, the following input r
 
 ### File system types
 
-Since command line applications often have to work with the file system, `FileInfo` and `DirectoryInfo` are supported. The following example shows the use of `FileInfo`, but it would work also with `DirectoryInfo`.
+Command line applications that work with the file system can use the <xref:System.IO.FileSystemInfo>, <xref:System.IO.FileInfo>, and <xref:System.IO.DirectoryInfo> types. The following example shows the use of `FileSystemInfo`:
```
Comment (gewarren) [inline docs/standard/commandline/model-binding.md:152]:
```suggestion
Command-line applications that work with the file system can use the <xref:System.IO.FileSystemInfo>, <xref:System.IO.FileInfo>, and <xref:System.IO.DirectoryInfo> types. The following example shows the use of `FileSystemInfo`:
```

File: docs/standard/commandline/define-commands.md [hunk 2]
```diff
@@ -146,6 +148,15 @@ Options:
   -?, -h, --help          Show help and usage information
 ```
 
+
+## Set argument arity
+
+You can explicitly set argument [arity](syntax.md#argument-arity) by using the `Arity` property, but in most cases that is not necessary. `System.CommandLine` automatically determines the argument arity based on the argument type:
+
+* `Boolean` arguments - `ZeroOrOne`
```
Comment (gewarren) [inline docs/standard/commandline/define-commands.md:156]: Are these enum fields or ? If so, maybe put this info in a table with a column header.

**Review by gewarren — APPROVED** [review]

File: docs/standard/commandline/define-commands.md [hunk 3]
```diff
@@ -132,6 +132,8 @@ If the command line for this example app doesn't include `--endpoint`, an error
 Option '--endpoint' is required.
 ```
 
+If a required option has a default value, the option doesn't have to be specified on the command line. In that case, the default value provides the required option value.
+
 ## Hidden commands, options, and arguments
 
 You might want to support a a command, option, or argument, but avoid making it easy to discover. For example, it might be a deprecated or administrative or preview feature. Use the `IsHidden` property to prevent users from discovering such features by using tab completion or help, as shown in the following example:
```
Comment (jonsequitur) [inline docs/standard/commandline/define-commands.md:139]: There's a duplicate `a` here: `You might want to support a a command`.

File: docs/standard/commandline/define-commands.md (hunk 2 above)
Comment (jonsequitur) [reply docs/standard/commandline/define-commands.md:155]: They are static members of the `ArgumentArity` type.

File: docs/standard/commandline/model-binding.md [hunk 4]
```diff
@@ -149,13 +149,17 @@ Because `AllowMultipleArgumentsPerToken` is set to `true`, the following input r
...
+:::code language="csharp" source="snippets/model-binding/csharp/Program.cs" id="filesysteminfo" :::
+
+With `FileInfo` and `DirectoryInfo` the pattern matching code is not required:
+
+:::code language="csharp" source="snippets/model-binding/csharp/Program.cs" id="filesysteminfo" :::
 
 ### Anything with a string constructor
 
-`FileInfo` and `DirectoryInfo` aren't special cases. Any type having a constructor that takes a single string parameter can be bound in this way. For example, the same code works with a `Uri` instead.
+The file system types aren't special cases. Any type having a constructor that takes a single string parameter can be bound in this way. For example, code that would work with `FileInfo` works with a <xref:System.Uri> instead.
```
Comment (jonsequitur) [inline docs/standard/commandline/model-binding.md:162]: This is no longer true after the work to support NativeAOT compilation, unfortunately. These are special cases. (`System.Uri` is another, and I expect the list might grow over time).

File: docs/standard/commandline/define-commands.md (hunk 2 above)
Comment (tdykstra) [reply docs/standard/commandline/define-commands.md:155]: Good idea. It's not an enum but it works like one.

File: docs/standard/commandline/model-binding.md (hunk 4 above)
Comment (tdykstra) [reply docs/standard/commandline/model-binding.md:162]: Are these now the only special cases or are there others that we should mention?

Comment (jonsequitur) [reply docs/standard/commandline/model-binding.md:162]: Technically, pretty much all types are now special cases. The two main conventions we used before doing the trimming work were the single-string constructor convention and the presence of a `TypeConverter`. Removing the latter also meant types like `DateTime` and `int` needed to be accounted for with special cases. You can see the implementations here: https://github.com/dotnet/command-line-api/blob/021ec68a4cb510c2cc125c6ebb78b9cfd4e3847a/src/System.CommandLine/Binding/ArgumentConverter.StringConverters.cs#L11

File: docs/standard/commandline/model-binding.md [hunk 5]
```diff
@@ -157,12 +157,30 @@ With `FileInfo` and `DirectoryInfo` the pattern matching code is not required:
...
 
-### Anything with a string constructor
+### Other supported types
 
-The file system types aren't special cases. Any type having a constructor that takes a single string parameter can be bound in this way. For example, code that would work with `FileInfo` works with a <xref:System.Uri> instead.
+Many types that have a constructor that takes a single string parameter can be bound in this way. For example, code that would work with `FileInfo` works with a <xref:System.Uri> instead.
 
 :::code language="csharp" source="snippets/model-binding/csharp/Program.cs" id="uri" :::
 
+Besides the file system types and `Uri`, the following types are supported:
```
Comment (jonsequitur) [inline docs/standard/commandline/model-binding.md:166]: I expect this list will continue to grow until we can find a generalized approach that's also supportable via NativeAOT compilation.

Comment (baronfel) [reply docs/standard/commandline/model-binding.md:166]: I suspect we'll end up at the static interfaces feature that the C# team is working on as a long-term, reflection-free solution. But that's just my guess.

**Pull request merged**